### DISCUSSION TITLE
Recommendations for slow broker startup

### DISCUSSION THREAD

**Discussion by PaulLiang1** [strimzi/strimzi-kafka-operator#4488] [General]
Hi, 

We recently have a few broker that suffered a unclean shutdown, and it could take quite a bit of time (~hour) for a broker to come back up;

Usually these are the log lines that shown during startup.
- `...kafka Loading producer state from snapshot file...` 
- `...kafka Recovering unflushed segment...`

Sometimes the this could exceed the duration of what liveness probe specifies, and cause the broker to stuck in a crash loop. 
Current workaround is to keep increasing the value of `initialDelaySeconds`.

Question: 
- Would increasing `terminationGracePeriodSeconds` help to reduce the likelihood of an unclean shutdown, so that there are less work required for broker startup hence making it faster?
- Is there a more elegant way of handling slow broker as setting `initialDelaySeconds` is more of a guess work. Would `startupProbe` more appropriate for such use case (even though Operator does not use this probe right now)? 

Thanks,
Paul

**Comment by scholzj** [comment]
You do not mention what Strimzi version you use. I think that if you use a reasonably recent version of Strimzi you should not need to change any initial delays. And yes, if you think this happened because your broker is too busy when shutting down, you can try to increase the grace period (you can do it your self in the `.spec.kafka.template` section of the Kafka custom resource). But from my experience these things usually happen when the broker is killed for other reasons without any proper shutdown at all (node crash, storage failure etc.). But you would need to check the logs to see what exactly was the cause.

**Reply by PaulLiang1** [reply]
Thanks for the reply! 

Strimizi version is `0.20.1` with Kafka `2.6.0` if that helps. 

Will give it a go and report back. Thanks!

**Reply by scholzj** [reply]
In 0.20 I would not expect you to need to play with any delays anymore. If you do need it, it would be great if you can provide your Kafka CR and the logs from the broker affected by this. I can try to see if I find some bug or something in there.

**Reply by PaulLiang1** [reply]
Please find the logs attached. 
Sorry I had to remove some information such as domain names / topic names etc. 

----

Both a short version and a long version of the log is attached. 
Short version removes some of the verbose log lines for brievity. 

From the logs line can be seen that
- `starting` msgs starts at `2021-03-03T02:38:19.632Z`
- `Terminating process due to signal SIGTERM` starts at `2021-03-03T02:50:58.223Z` 
- which is slightly higher than the `initialDelaySeconds` set with `600s` + actual time takes for failureThresholds
- Something that puzzles me is that the kill reason is 'Readiness probe failed: '  instead of liveness probe, as I thought readiness probe would only stop traffic from routing to the pods. 


Env:
- `k8s version`: `v1.16.15-eks-ad4801`
- `Strimizi version`: `0.20.1` 
- `Kafka`:  `2.6.0`


TLDR version
```
Starting

...
...Recovering unflushed segment...
...Loading producer state from snapshot file...
...Completed load of Log...
...

Terminating process due to signal SIGTERM
shutting down
Kafka server is still starting up, cannot shut down
Halting Kafka.
```


kube event that kills the pod
```
┌──────────────────────────────────────────────────────────────────────────────────────────── YAML(kafka/kafka-cluster-kafka-3.1668b5370d6e18b3) ────────────────────────────────────────────────────────────────────────────────────────────┐
│ apiVersion: v1                                                                                                                                                                                                                             │
│ count: 3                                                                                                                                                                                                                                   │
│ eventTime: null                                                                                                                                                                                                                            │
│ firstTimestamp: "2021-03-03T02:48:34Z"                                                                                                                                                                                                     │
│ involvedObject:                                                                                                                                                                                                                            ││   apiVersion: v1                                                                                                                                                                                                                           │
│   fieldPath: spec.containers{kafka}                                                                                                                                                                                                        │
│   kind: Pod                                                                                                                                                                                                                                │
│   name: kafka-cluster-kafka-3                                                                                                                                                                                                              │
│   namespace: kafka                                                                                                                                                                                                                         │
│   resourceVersion: "77017782"                                                                                                                                                                                                              │
│   uid: 504148c7-1fa9-4eed-9906-72dafb24beed                                                                                                                                                                                                │
│ kind: Event                                                                                                                                                                                                                                │
│ lastTimestamp: "2021-03-03T02:50:34Z"                                                                                                                                                                                                      │
│ message: 'Readiness probe failed: '                                                                                                                                                                                                        │
│ metadata:                                                                                                                                                                                                                                  │
│   creationTimestamp: "2021-03-03T02:48:34Z"                                                                                                                                                                                                │
│   name: kafka-cluster-kafka-3.1668b5370d6e18b3                                                                                                                                                                                             │
│   namespace: kafka                                                                                                                                                                                                                         │
│   resourceVersion: "77022685"                                                                                                                                                                                                              │
│   selfLink: /api/v1/namespaces/kafka/events/kafka-cluster-kafka-3.1668b5370d6e18b3                                                                                                                                                         │
│   uid: 409a66ed-1a4e-4652-a096-b39b6e8d704d                                                                                                                                                                                                │
│ reason: Unhealthy                                                                                                                                                                                                                          │
│ reportingComponent: ""                                                                                                                                                                                                                     │
│ reportingInstance: ""                                                                                                                                                                                                                      │
│ source:                                                                                                                                                                                                                                    │
│   component: kubelet                                                                                                                                                                                                                       │
│   host: XXXX                                                                                                                                                                                       │
│ type: Warning                                                                                                                                                                                                                              │
│
```

```
┌────────────────────────────────────────────────────────────────────────────────────────── Describe(kafka/kafka-cluster-kafka-3.1668b55884c61008) ──────────────────────────────────────────────────────────────────────────────────────────┐
│ Name:             kafka-cluster-kafka-3.1668b55884c61008                                                                                                                                                                                   │
│ Namespace:        kafka                                                                                                                                                                                                                    │
│ Labels:           <none>                                                                                                                                                                                                                   │
│ Annotations:      <none>                                                                                                                                                                                                                   │
│ API Version:      v1                                                                                                                                                                                                                       ││ Count:            1                                                                                                                                                                                                                        │
│ Event Time:       <nil>                                                                                                                                                                                                                    │
│ First Timestamp:  2021-03-03T02:50:58Z                                                                                                                                                                                                     │
│ Involved Object:                                                                                                                                                                                                                           │
│   API Version:       v1                                                                                                                                                                                                                    │
│   Field Path:        spec.containers{kafka}                                                                                                                                                                                                │
│   Kind:              Pod                                                                                                                                                                                                                   │
│   Name:              kafka-cluster-kafka-3                                                                                                                                                                                                 │
│   Namespace:         kafka                                                                                                                                                                                                                 │
│   Resource Version:  77017782                                                                                                                                                                                                              │
│   UID:               504148c7-1fa9-4eed-9906-72dafb24beed                                                                                                                                                                                  │
│ Kind:                Event                                                                                                                                                                                                                 │
│ Last Timestamp:      2021-03-03T02:50:58Z                                                                                                                                                                                                  │
│ Message:             Stopping container kafka                                                                                                                                                                                              │
│ Metadata:                                                                                                                                                                                                                                  │
│   Creation Timestamp:  2021-03-03T02:50:58Z                                                                                                                                                                                                │
│   Resource Version:    77022822                                                                                                                                                                                                            │
│   Self Link:           /api/v1/namespaces/kafka/events/kafka-cluster-kafka-3.1668b55884c61008                                                                                                                                              │
│   UID:                 bdeca284-ddce-4b17-8738-986d2a3c4a35                                                                                                                                                                                │
│ Reason:                Killing                                                                                                                                                                                                             │
│ Reporting Component:                                                                                                                                                                                                                       │
│ Reporting Instance:                                                                                                                                                                                                                        │
│ Source:                                                                                                                                                                                                                                    │
│   Component:  kubelet                                                                                                                                                                                                                      │
│   Host:       XXX                                                                                                                                                                                 │
│ Type:         Normal                                                                                                                                                                                                                       │
│ Events:       <none>
```

----
CRD:
[crd.yaml.txt](https://github.com/strimzi/strimzi-kafka-operator/files/6073615/crd.yaml.txt)

Log (Slim):
[slim.log](https://github.com/strimzi/strimzi-kafka-operator/files/6073620/slim.log)

Log (Longer version):
[longer.log.gz](https://github.com/strimzi/strimzi-kafka-operator/files/6073618/longer.log.gz)

**Reply by scholzj** [reply]
Can you share the Kafka CR as well?

**Reply by PaulLiang1** [reply]
Hi @scholzj , 

For CR I presume you mean the resource def for `Kafka@kafka.strimzi.io/v1beta1`?
If so, I made a typo above.. `crd.yaml.txt` it's actually the CR, not CRD...

**Reply by forsberg** [reply]
@scholzj - can you elaborate on what changes in Strimzi 0.20 eliminate the need to change readyness delays? 

Had some similar trouble on a Strimzi 0.19/Kafka 2.5.0 cluster where one node got stuck not being able to start, taking a very long time to load its log. I increased the initialDelaySeconds setting, but somehow got into a situation where the cluster operator force-rolled my pod even before the initial delay seconds had passed.

I had to turn off the cluster operator and wait for the pod to come online to solve this. I guess all of this is fixed in a later Strimzi version.

I also increased `num.recovery.threads.per.data.dir` from its default 1 to 8, which seems to have helped in increasing the read bandwidth used to read from the AWS EC2 st1 volumes backing the PVC, decreasing startup time.

**Reply by scholzj** [reply]
@forsberg I do not think there was anything specific in 0.20. I would expect it to behave the same way with 0.19. The readiness / liveness probes deal with the pod startup - they do not deal with the rolling updates from the operator. So that is anyway unrelated - @tombentley and @sknot-rh worked on the Kafka Roller ... they might know more when / why would it force-roll the pod.

**Reply by forsberg** [reply]
@scholzj Aha. I was doing multiple changes to the Kafka CR, and did not necessarily let the cluster stabilize between attempts, so the operator was probably very confused. I see in code in git it adds an explanation on why it would force-roll, something 0.19 lacks.

After getting the whole cluster stable again I tried manually killing the troublesome pod again, and it seems it needs more than the default 30s to flush all its segments, because again it's taking a very long time to start, even after a controlled shutdown. So I will try increasing `terminationGracePeriodSeconds` to see if this helps. Behaviour seems to vary with pod, might depend on noisy neighbour on Kubernetes node or similar. 

It is a bit of a tricky situation. The pod is "kind of ready". I'm pondering if it would be better to:

* Modify the readynessProbe script to mark a node that is recovering unflushed segments as ready. Don't know how tricky it is to detect this state.
* Let the Cluster operator do what it already does - verify in ISR status for all partitions to determine when it can roll the next pod in the statefulset.

**Reply by forsberg** [reply]
I should mention this is my second Strimzi on AWS EKS cluster, this second one being at a different AWS region (and different job :) ) than the first one where I'm having this problem.

Being able to configure a [startup probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-startup-probes) could be another way forward.

**Reply by scholzj** [reply]
The startup probe basically just postpones the readiness probe. So it is not really something what solves this issue on its own. It just adds more timeouts to play with. The terminationGracePeriod is a bit tricky because you do not easily see if it is enough or not. We just had a discussion about it the other day, but without some good conclusion TBH.

**Reply by forsberg** [reply]
Although I agree that timeouts are usually just an indication that the problem has not been solved properly, if I compare the following two options:

* readinessProbe with 20 minutes of initial delay, and periodSeconds + failureThreshold that makes sense during normal operations to detect deadlocks etc.
* startupProbe with periodSeconds=10 and failureThreshold=120

..the latter has the advantage that it will mark the pod ready much faster during nominal startup, i.e when shutdown was clean, and I could keep readinessProbe at default values where it detects abnormal situations while the pod is running.

**Reply by danjo133** [reply]
> I should mention this is my second Strimzi on AWS EKS cluster, this second one being at a different AWS region (and different job :) ) than the first one where I'm having this problem.
> 

Yes, your old sins are now haunting me ;) I should follow your comments on github so I get a heads up for the upgrades I'm doing!

> Being able to configure a [startup probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-startup-probes) could be another way forward.

I also stopped the operator to let the pods come alive (some 15+ minutes per pod). I saw various configurable delays, but am a bit unsure on what the best strategy is.

**Reply by jonerer** [reply]
I'm looking at the termination logs a cluster exhibiting this issue. Strimzi 0.22. The final lines in the shutdown operation seem to indicate an OK shutdown:

```
2022-01-13 14:20:27 +01:00 | production-kafka-2
2022-01-13 13:20:27,810 INFO [KafkaServer id=2] shut down completed (kafka.server.KafkaServer) [kafka-shutdown-hook]
2022-01-13 14:20:27 +01:00 | production-kafka-2
2022-01-13 13:20:27,805 INFO [SocketServer brokerId=2] Shutdown completed (kafka.network.SocketServer) [kafka-shutdown-hook]
2022-01-13 14:20:27 +01:00 | production-kafka-2
2022-01-13 13:20:27,792 INFO [SocketServer brokerId=2] Shutting down socket server (kafka.network.SocketServer) [kafka-shutdown-hook]
2022-01-13 14:20:27 +01:00 | production-kafka-2
2022-01-13 13:20:27,791 INFO [ThrottledChannelReaper-Request]: Shutdown completed (kafka.server.ClientQuotaManager$ThrottledChannelReaper) [kafka-shutdown-hook]
2022-01-13 14:20:27 +01:00 | production-kafka-2
2022-01-13 13:20:27,791 INFO [ThrottledChannelReaper-Request]: Stopped (kafka.server.ClientQuotaManager$ThrottledChannelReaper) [ThrottledChannelReaper-Request]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,812 INFO [ThrottledChannelReaper-Produce]: Stopped (kafka.server.ClientQuotaManager$ThrottledChannelReaper) [ThrottledChannelReaper-Produce]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,812 INFO [ThrottledChannelReaper-Request]: Shutting down (kafka.server.ClientQuotaManager$ThrottledChannelReaper) [kafka-shutdown-hook]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,812 INFO [ThrottledChannelReaper-Produce]: Shutdown completed (kafka.server.ClientQuotaManager$ThrottledChannelReaper) [kafka-shutdown-hook]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,646 INFO [ThrottledChannelReaper-Fetch]: Shutdown completed (kafka.server.ClientQuotaManager$ThrottledChannelReaper) [kafka-shutdown-hook]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,646 INFO [ThrottledChannelReaper-Produce]: Shutting down (kafka.server.ClientQuotaManager$ThrottledChannelReaper) [kafka-shutdown-hook]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,646 INFO [ThrottledChannelReaper-Fetch]: Stopped (kafka.server.ClientQuotaManager$ThrottledChannelReaper) [ThrottledChannelReaper-Fetch]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,216 INFO [ThrottledChannelReaper-Fetch]: Shutting down (kafka.server.ClientQuotaManager$ThrottledChannelReaper) [kafka-shutdown-hook]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,215 INFO Session: 0x201f854863b03f6 closed (org.apache.zookeeper.ZooKeeper) [kafka-shutdown-hook]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,215 INFO channel is disconnected: [id: 0xc148bf50, L:/10.0.91.75:40568 ! R:production-zookeeper-client/172.20.22.45:2181] (org.apache.zookeeper.ClientCnxnSocketNetty) [nioEventLoopGroup-2-1]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,215 INFO EventThread shut down for session: 0x201f854863b03f6 (org.apache.zookeeper.ClientCnxn) [main-EventThread]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,215 INFO [ZooKeeperClient Kafka server] Closed. (kafka.zookeeper.ZooKeeperClient) [kafka-shutdown-hook]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,215 INFO channel is told closing (org.apache.zookeeper.ClientCnxnSocketNetty) [nioEventLoopGroup-2-1]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,214 INFO channel is told closing (org.apache.zookeeper.ClientCnxnSocketNetty) [kafka-shutdown-hook]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,211 INFO [ZooKeeperClient Kafka server] Closing. (kafka.zookeeper.ZooKeeperClient) [kafka-shutdown-hook]
2022-01-13 14:20:26 +01:00 | production-kafka-2
2022-01-13 13:20:26,209 INFO [ReplicaStateMachine controllerId=2] Stopped replica state machine (kafka.controller.ZkReplicaStateMachine) [kafka-shutdown-hook]
```

etc... But the startup still takes ~20 minutes. Currently investigating whether our nodes are starved for bandwith to EBS or we're limited by disk type (currently using st1 on AWS, as indicated by https://aws.amazon.com/blogs/big-data/best-practices-for-running-apache-kafka-on-aws/)

The data folder for each broker is ~450GB. Is 20 minutes of startup time reasonable, given that shutdown was clean?

Also: Hello again @forsberg ! Did you get any further?
